qualtran/bloqs/gf_arithmetic/gf2_inverse.py

```diff
@@ -129,90 +149,106 @@ def my_static_costs(self, cost_key: 'CostKey'):
 
         return NotImplemented
 
+    @cached_property
+    def _bits(self) -> list[int]:
+        k1 = bit_length(self.bitsize - 1) - 1
+        return [-1] + [k1 - i for i, b in enumerate(np.binary_repr(self.bitsize - 1)) if b == '1']
+
     def build_composite_bloq(self, bb: 'BloqBuilder', *, x: 'Soquet') -> Dict[str, 'SoquetT']:
         if is_symbolic(self.bitsize):
             raise DecomposeTypeError(f"Cannot decompose symbolic {self}")
 
-        result = bb.allocate(dtype=self.qgf)
         if self.bitsize == 1:
-            x, result = bb.add(GF2Addition(self.bitsize), x=x, y=result)
+            result = bb.allocate(dtype=self.qgf)
+            x, result = bb.add(GF2Addition(self.qgf), x=x, y=result)
             return {'x': x, 'result': result}
 
-        junk = []
-        beta = x
-        is_first = True
-        bitsize_minus_one = int(self.bitsize - 1)
-        n_iters = bitsize_minus_one.bit_length()
-        for i in range(n_iters):
-            if (1 << i) & bitsize_minus_one:
-                if is_first:
-                    beta, result = bb.add(GF2Addition(self.bitsize), x=beta, y=result)
-                    is_first = False
-                else:
-                    for j in range(2**i):
-                        result = bb.add(GF2Square(self.bitsize), x=result)
-                    beta, result, new_result = bb.add(
-                        GF2MulViaKaratsuba(self.bitsize), x=beta, y=result
-                    )
-                    junk.append(result)
-                    result = new_result
-            if i != n_iters - 1:
-                beta_squared = bb.allocate(dtype=self.qgf)
-                beta, beta_squared = bb.add(GF2Addition(self.bitsize), x=beta, y=beta_squared)
-                for j in range(2**i):
-                    beta_squared = bb.add(GF2Square(self.bitsize), x=beta_squared)
-                beta, beta_squared, beta_new = bb.add(
-                    GF2MulViaKaratsuba(self.bitsize), x=beta, y=beta_squared
-                )
-                junk.extend([beta, beta_squared])
-                beta = beta_new
-        junk.append(beta)
-        result = bb.add(GF2Square(self.bitsize), x=result)
-        x = junk.pop(0)
-        assert len(junk) == self.n_junk_regs, f'{len(junk)=}, {self.n_junk_regs=}'
-        return {'x': x, 'result': result, 'junk': np.array(junk)}
+        t = (self.bitsize - 1).bit_count()
+        k1 = bit_length(self.bitsize - 1) - 1
+        k = max(k1 + t - 1, k1 + 1)
+        f = [x] + [None] * k
+        f[k] = bb.allocate(self.bitsize, self.qgf)
+        f = cast(list['Soquet'], f)
+        for i in range(1, k1 + 1):
+            f[i - 1], f[k] = bb.add(GF2Addition(self.qgf), x=f[i - 1], y=f[k])
+            f[k] = bb.add(GF2Square(self.qgf, 2 ** (i - 1)), x=f[k])
+            f[i - 1], f[k], f[i] = bb.add(GF2MulViaKaratsuba(self.qgf), x=f[i - 1], y=f[k])
```

Comment:
One comment - Can you make `GF2MulViaKaratsuba(self.qgf)` a property on the class such that in the decomposition you call `self.gf2_mul` ? This will be helpful because if I need to use a different multiplier, I can just derive from this class and update the `gf_mul` property to return a more optimized multiplier with the same signature.